### PR TITLE
Simplify INI-style config reader using C++11 STL

### PR DESCRIPTION
**Pull request by fuhaoda**
Use C++11 STL to simplify the   "./src/common/config.h"  file.

### CODE REVIEW

**Comment by trivialfis** [comment]
@fuhaoda Hi, any update on this?  And could you please add some simple tests?

**Comment by fuhaoda** [comment]
> @fuhaoda Hi, any update on this? And could you please add some simple tests?

Is it only a format issue? I don't know what's wrong here. It can be compiled and pass test on my Mac, and I compared the results and test case. All are the same before I submit the pull request.

**Comment by trivialfis** [comment]
@fuhaoda It has some linter issues on Linux, and somehow failed tests on Windows.   Would you like to fix the linter first so that we can see how's it doing on Linux?  See: https://xgboost-ci.net/blue/organizations/jenkins/xgboost/detail/PR-4478/1/pipeline

and look for some lines like this:

> [2019-05-20T02:50:11.138Z] src/cli_main.cc:345:  Missing spaces around =  [whitespace/operators] [4]
> [2019-05-20T02:50:11.138Z] src/common/config.h:17:  Do not indent within a namespace  [runtime/indentation_namespace] [4]
> [2019-05-20T02:50:11.138Z] src/common/config.h:21:  Do not indent within a namespace  [runtime/indentation_namespace] [4]

**Comment by fuhaoda** [comment]
@trivialfis Thank you so much! Trying really hard and fixed the linter issues.  Could you please take a look on what's the issue on Windows test? My guess is the path issue to include character like \.

**Comment by hcho3** [comment]
@fuhaoda You were using older versions of the submodules `dmlc-core` and `rabit`.

File: src/common/config.h [hunk 1]
```diff
@@ -12,181 +12,74 @@
...
+    char comment = '#';
+    std::string line{};
+    std::string name{};
+    std::string value{};
 
- private:
-  int ch_buf_;
-  std::string s_name_, s_val_, s_buf_;
+    while (!fi_.eof()) {
+      getline(fi_, line);  // read a line of configure file
```
Comment (hcho3) [inline src/common/config.h]:
```suggestion
      std::getline(fi_, line);  // read a line of configure file
```

**Comment by fuhaoda** [comment]
@hcho3 any thoughts on the check failures? it was successful yesterday...

**Comment by hcho3** [comment]
@fuhaoda Looks like it's out of memory. Let me re-start the test again.

File: src/common/config.h [hunk 2]
```diff
@@ -12,181 +12,74 @@
...
-    return fin_.eof();
+  ~ConfigParse() {
+    fi_.close();
   }
 
  private:
-  std::istream &fin_;
-};
+  std::ifstream fi_;
+  std::string allowableChar =
```
Comment (trivialfis) [inline src/common/config.h]: I'm curious why clang-tidy didn't catch this styling error.

File: src/common/config.h [hunk 3]
```diff
@@ -1,8 +1,8 @@
 /*!
- * Copyright 2014 by Contributors
+ * Copyright 2019 by Contributors
```
Comment (trivialfis) [inline src/common/config.h]: 2014-2019

File: src/common/config.h (hunk 2 above)
Comment (hcho3) [reply src/common/config.h]: @trivialfis I don't see any issue with this line?

**Comment by trivialfis** [comment]
@fuhaoda Sorry for the late reply.  I can't push to your branch, so could you apply the following patch and I will merge?
[0001-Fix-header-and-format.zip](https://github.com/dmlc/xgboost/files/3235447/0001-Fix-header-and-format.zip)

File: src/common/config.h (hunk 2 above)
Comment (trivialfis) [reply src/common/config.h]: @hcho3 private member suffix '_'.  I tried clang-tidy-8 and it didn't catch it either.

**Comment by fuhaoda** [comment]
@trivialfis Thanks a lot. Just followed your suggestion and changed them all.

**Comment by fuhaoda** [comment]
@hcho3 is it an out of memory issue again? could you please re-start the test again? looking forward to the final merge.

**Comment by hcho3** [comment]
@fuhaoda No, this time it was compilation failure, because the `cub` submodule was pointing at wrong commit. I fixed the problem

**Comment by fuhaoda** [comment]
@hcho3 likely to need your help again : ) failed in a different test

**Comment by trivialfis** [comment]
@hcho3 I will leave it for you to merge it.  Thanks for all the helps here.

**Comment by hcho3** [comment]
@trivialfis I'm still looking for ways to eliminate out-of-memory errors that pop up once in a while.

**Pull request merged**